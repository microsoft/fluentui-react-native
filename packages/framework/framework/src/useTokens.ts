import { buildUseTokens as buildUseTokensCore, UseTokens as UseTokensCore } from '@fluentui-react-native/use-tokens';
import { themeHelper } from './themeHelper';
import { Theme } from '@fluentui-react-native/theme-types';
import { TokenSettings } from './useStyling';

export { applyTokenLayers, applyPropsToTokens, customizable, patchTokens } from '@fluentui-react-native/use-tokens';

export type UseTokens<TTokens> = UseTokensCore<TTokens, Theme>;

export function buildUseTokens<TTokens>(...tokens: TokenSettings<TTokens>[]): UseTokens<TTokens> {
  return buildUseTokensCore(themeHelper.getComponentInfo, ...tokens);
}
