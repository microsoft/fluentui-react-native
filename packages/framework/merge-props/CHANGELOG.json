{
  "name": "@fluentui-react-native/merge-props",
  "entries": [
    {
      "date": "Fri, 25 Sep 2020 19:21:43 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.3",
      "version": "0.1.3",
      "comments": {
        "patch": [
          {
            "comment": "Update react-native-win32 versions - enable logbox",
            "author": "acoates-ms@noreply.github.com",
            "commit": "a95ab1b66822cceacc74e24094f85c6917f58086",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    },
    {
      "date": "Wed, 23 Sep 2020 18:31:48 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.2",
      "version": "0.1.2",
      "comments": {
        "patch": [
          {
            "comment": "start publishing src to fix customer source maps",
            "author": "jasonmo@microsoft.com",
            "commit": "b3d7b5ba710bc581a6f4db91c91e4ffc185fc427",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    },
    {
      "date": "Fri, 11 Sep 2020 20:25:42 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "none": [
          {
            "comment": "Add samples and some README text to use-slots and use-styling packages",
            "author": "jasonmo@microsoft.com",
            "commit": "2566347850eff4ed67bfad2a85134481e088cba6",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    },
    {
      "date": "Fri, 28 Aug 2020 05:03:51 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "none": [
          {
            "comment": "only allow menu container to focus on mount",
            "author": "lehon@microsoft.com",
            "commit": "9c93388527af2ea3ba60336c56a3da0ce65e3688",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    },
    {
      "date": "Thu, 20 Aug 2020 00:14:01 GMT",
      "tag": "@fluentui-react-native/merge-props_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "split mergeProps and mergeStyles out of foundation-settings",
            "author": "jasonmo@microsoft.com",
            "commit": "f75e45ede85b8ebc6cd11cc66fbb1baedaea39b9",
            "package": "@fluentui-react-native/merge-props"
          }
        ]
      }
    }
  ]
}
