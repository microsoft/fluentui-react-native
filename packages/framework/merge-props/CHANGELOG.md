# Change Log - @fluentui-react-native/merge-props

This log was last generated on Fri, 25 Sep 2020 19:21:43 GMT and should not be manually modified.

<!-- Start content -->

## 0.1.3

Fri, 25 Sep 2020 19:21:43 GMT

### Patches

- Update react-native-win32 versions - enable logbox (acoates-ms@noreply.github.com)

## 0.1.2

Wed, 23 Sep 2020 18:31:48 GMT

### Patches

- start publishing src to fix customer source maps (jasonmo@microsoft.com)

## 0.1.1

Thu, 20 Aug 2020 00:14:01 GMT

### Patches

- split mergeProps and mergeStyles out of foundation-settings (jasonmo@microsoft.com)
