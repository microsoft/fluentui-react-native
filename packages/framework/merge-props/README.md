# `@fluentui-react-native/merge-props`

This package is now a pass-through package which references `@fluentui-react-native/framework-base` for backwards compatibility.

The current documentation can be found [here](../../framework-base/src/merge-props/README.md).
