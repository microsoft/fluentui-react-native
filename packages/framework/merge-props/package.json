{
  "name": "@fluentui-react-native/merge-props",
  "version": "0.8.0",
  "description": "Utility for merging props with styles and caching style combinations",
  "repository": {
    "type": "git",
    "url": "https://github.com/microsoft/fluentui-react-native",
    "directory": "packages/framework/merge-props"
  },
  "main": "src/index.ts",
  "module": "src/index.ts",
  "typings": "lib/index.d.ts",
  "onPublish": {
    "main": "lib-commonjs/index.js",
    "module": "lib/index.js"
  },
  "scripts": {
    "build": "fluentui-scripts build",
    "just": "fluentui-scripts",
    "clean": "fluentui-scripts clean",
    "depcheck": "fluentui-scripts depcheck",
    "test": "fluentui-scripts jest",
    "lint": "fluentui-scripts eslint",
    "start": "fluentui-scripts dev",
    "start-test": "fluentui-scripts jest-watch",
    "update-snapshots": "fluentui-scripts jest -u",
    "prettier": "fluentui-scripts prettier",
    "prettier-fix": "fluentui-scripts prettier --fix true"
  },
  "keywords": [],
  "author": "",
  "license": "MIT",
  "dependencies": {
    "@fluentui-react-native/immutable-merge": "^1.2.1",
    "@fluentui-react-native/memo-cache": "^1.3.1",
    "tslib": "^2.3.1"
  },
  "devDependencies": {
    "@fluentui-react-native/eslint-config-rules": "^0.1.1",
    "@fluentui-react-native/scripts": "^0.1.1",
    "@react-native/metro-config": "^0.72.0",
    "@types/jest": "^29.0.0",
    "react": "18.2.0",
    "react-native": "^0.72.0"
  },
  "peerDependencies": {
    "react": "18.2.0",
    "react-native": "^0.72.0"
  },
  "rnx-kit": {
    "kitType": "library",
    "alignDeps": {
      "presets": [
        "microsoft/react-native"
      ],
      "requirements": [
        "react-native@0.72"
      ],
      "capabilities": [
        "core",
        "core-android",
        "core-ios"
      ]
    }
  }
}
