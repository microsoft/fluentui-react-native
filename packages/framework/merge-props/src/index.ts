export { StyleProp } from './mergeStyles.types';
export { mergeStyles } from './mergeStyles';
export { mergeProps } from './mergeProps';
