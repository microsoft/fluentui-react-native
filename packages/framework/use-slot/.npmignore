node_modules
.gitignore
.gitattributes
.editorconfig
config.js
jest.config.js
tslint.json
tsconfig.json
jsconfig.json
webpack.config.js
webpack.serve.config.js
*.build.log
