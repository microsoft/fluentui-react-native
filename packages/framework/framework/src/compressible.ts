import { Theme } from '@fluentui-react-native/theme-types';
import { stagedComponent, StagedRender } from '@fluentui-react-native/use-slot';
import { CustomizableComponent } from '@fluentui-react-native/use-tokens';
import { UseTokens } from './useTokens';
import { TokenSettings } from './useStyling';

export function compressible<TProps, TTokens>(
  fn: StagedRender<TProps>,
  useTokens: UseTokens<TTokens>,
): CustomizableComponent<TProps, TTokens, Theme> {
  type ThisComponent = CustomizableComponent<TProps, TTokens, Theme>;

  const injectedWrapper = (props: TProps) => fn(props, useTokens);
  const component = stagedComponent(injectedWrapper) as ThisComponent;

  component.customize = (...tokens: TokenSettings<TTokens>[]) => {
    const useTokensNew = useTokens.customize(...tokens);
    return compressible(fn, useTokensNew);
  };

  return component;
}
