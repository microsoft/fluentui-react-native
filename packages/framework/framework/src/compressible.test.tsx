/** @jsx withSlots */
import { buildUseTokens, UseTokens } from './useTokens';
import { applyTokenLayers } from '@fluentui-react-native/use-tokens';
import { TextProps, TextStyle, Text, View } from 'react-native';
import { compressible } from './compressible';
import { Theme } from '@fluentui-react-native/theme-types';
import { mergeStyles } from '@fluentui-react-native/merge-props';
import { useSlot, withSlots } from '@fluentui-react-native/use-slot';
import * as React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';

type Variant = 'normal' | 'header' | 'caption';
interface VariantTextTokens {
  color?: string;
  fontSize?: TextStyle['fontSize'];
  fontWeight?: TextStyle['fontWeight'];
  defaultType?: Variant;
  header?: VariantTextTokens;
  caption?: VariantTextTokens;
}

type VariantTextProps = TextProps & { variant?: Variant };

const useVariantTokens = buildUseTokens<VariantTextTokens>({
  color: 'black',
  fontSize: 12,
  fontWeight: '500',
  defaultType: 'normal',
  header: {
    color: 'blue',
    fontSize: 16,
    fontWeight: '700',
  },
  caption: {
    color: 'gray',
    fontWeight: '300',
  },
});

const VariantText = compressible<VariantTextProps, VariantTextTokens>(
  (props: VariantTextProps, useTokens: UseTokens<VariantTextTokens>) => {
    // fake theme here
    const theme: Theme = {} as Theme;
    // get the tokens
    let [tokens, cache] = useTokens(theme);
    // split the props, defaulting the variant type to the default type from the tokens
    const { variant = tokens.defaultType, style, ...rest } = props;
    // now apply the alternate layer tokens as appropriate
    [tokens, cache] = applyTokenLayers(tokens, ['normal', 'header', 'caption'], cache, (layer) => layer === variant);
    // merge styles together with what is passed in
    const mergedStyle = mergeStyles({ color: tokens.color, fontSize: tokens.fontSize, fontWeight: tokens.fontWeight }, style);
    // now get the slot
    const InnerText = useSlot<TextProps>(Text, { ...rest, style: mergedStyle });

    return (extra: VariantTextProps, children: React.ReactNode) => {
      return <InnerText {...extra}>{children}</InnerText>;
    };
  },
  useVariantTokens,
);

const SuperHeader = VariantText.customize({ header: { fontSize: 24, color: 'purple' } });

type LabelProps = {
  headerText: string;
  captionText?: string;
  headerSlot?: React.FunctionComponent<VariantTextProps>;
  captionSlot?: React.FunctionComponent<VariantTextProps>;
};

type LabelTokens = {
  headerVariant?: Variant;
  captionVariant?: Variant;
};

const useLabelTokens = buildUseTokens<LabelTokens>({
  headerVariant: 'header',
  captionVariant: 'caption',
});

const Label = compressible<LabelProps, LabelTokens>((props: LabelProps, useTokens: UseTokens<LabelTokens>) => {
  const theme: Theme = {} as Theme;
  const [tokens] = useTokens(theme);
  const { headerText, captionText, headerSlot, captionSlot } = props;
  const Header = useSlot<VariantTextProps>(headerSlot || VariantText, { variant: tokens.headerVariant });
  const Caption = useSlot<VariantTextProps>(captionSlot || VariantText, { variant: tokens.captionVariant });
  return () => {
    if (captionText) {
      return (
        <View>
          <Header>{headerText}</Header>
          <Caption>{captionText}</Caption>
        </View>
      );
    }
    return <Header>{headerText}</Header>;
  };
}, useLabelTokens);

/**
 * this wrapper solves the (so-far) inexplicable type errors from the matchers in typescript
 */
function snapshotTestTree(tree: any) {
  (expect(toJson(tree)) as any).toMatchSnapshot();
}

describe('compressible tests', () => {
  /** first render the component with no updates */
  it('Two labels, one with caption and one without', () => {
    const tree = mount(
      <View>
        <Label headerText="Header1" />
        <Label headerText="Header2" captionText="Caption2" />
      </View>,
    );
    snapshotTestTree(tree);
  });

  it('Two labels, one plugging in SuperHeader instead', () => {
    const tree = mount(
      <View>
        <Label headerText="Super Header" headerSlot={SuperHeader} captionText="Normal caption" />
        <Label headerText="Normal Header" captionText="Another normal caption" />
      </View>,
    );
    snapshotTestTree(tree);
  });
});
