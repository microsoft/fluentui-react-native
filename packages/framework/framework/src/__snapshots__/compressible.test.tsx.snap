// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compressible tests Two labels, one plugging in SuperHeader instead 1`] = `
<View>
  <View>
    <component
      captionText="Normal caption"
      headerSlot={[Function]}
      headerText="Super Header"
    >
      <View>
        <View>
          <Text
            style={
              Object {
                "color": "purple",
                "fontSize": 24,
                "fontWeight": "700",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "purple",
                  "fontSize": 24,
                  "fontWeight": "700",
                }
              }
            >
              Super Header
            </Text>
          </Text>
          <Text
            style={
              Object {
                "color": "gray",
                "fontSize": 12,
                "fontWeight": "300",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "gray",
                  "fontSize": 12,
                  "fontWeight": "300",
                }
              }
            >
              Normal caption
            </Text>
          </Text>
        </View>
      </View>
    </component>
    <component
      captionText="Another normal caption"
      headerText="Normal Header"
    >
      <View>
        <View>
          <Text
            style={
              Object {
                "color": "blue",
                "fontSize": 16,
                "fontWeight": "700",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "blue",
                  "fontSize": 16,
                  "fontWeight": "700",
                }
              }
            >
              Normal Header
            </Text>
          </Text>
          <Text
            style={
              Object {
                "color": "gray",
                "fontSize": 12,
                "fontWeight": "300",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "gray",
                  "fontSize": 12,
                  "fontWeight": "300",
                }
              }
            >
              Another normal caption
            </Text>
          </Text>
        </View>
      </View>
    </component>
  </View>
</View>
`;

exports[`compressible tests Two labels, one with caption and one without 1`] = `
<View>
  <View>
    <component
      headerText="Header1"
    >
      <Text
        style={
          Object {
            "color": "blue",
            "fontSize": 16,
            "fontWeight": "700",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "blue",
              "fontSize": 16,
              "fontWeight": "700",
            }
          }
        >
          Header1
        </Text>
      </Text>
    </component>
    <component
      captionText="Caption2"
      headerText="Header2"
    >
      <View>
        <View>
          <Text
            style={
              Object {
                "color": "blue",
                "fontSize": 16,
                "fontWeight": "700",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "blue",
                  "fontSize": 16,
                  "fontWeight": "700",
                }
              }
            >
              Header2
            </Text>
          </Text>
          <Text
            style={
              Object {
                "color": "gray",
                "fontSize": 12,
                "fontWeight": "300",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "gray",
                  "fontSize": 12,
                  "fontWeight": "300",
                }
              }
            >
              Caption2
            </Text>
          </Text>
        </View>
      </View>
    </component>
  </View>
</View>
`;
